### PR TITLE
Add redirection for join section

### LINKED ISSUE #424: Issue: Not redirecting to join section page
# Issue: Not redirecting to join section page

**Description**:

After submitting the signup form, the user is not redirected to the join section page.

**Expected behaviour**:

After submitting the signup form, the user should be redirected to the join section page.

### PR DESCRIPTION
**Pull request by sahsisunny**
# Title : Add redirect to join section page after signup form submission

### What is the change?

- `join.html`  : Remove "Follow to Ankush" text from join section.
- `js/join.js` : Add code to set `hasJoinVisited` value to localStorage.
- `js/goto.js` : Add the redirect to join section page after signup form submission(with check for `hasJoinVisited` value)

### Is Development Tested?

- [x] Yes
- [ ] No

### After Change Screenshots

image
### Video After Changes - **Sign in with GitHub**

https://user-images.githubusercontent.com/70854507/223870063-902612ad-9040-4120-8a23-5d982dda2e58.mp4

### Video After Changes - **Direct realdevsquad.com/join**

https://user-images.githubusercontent.com/70854507/223870187-b4a57929-b8a8-48f3-9ed9-74727c905092.mp4

### CODE REVIEW

File: js/goto.js [hunk 1]
```diff
@@ -1,12 +1,24 @@
...
+      : page == 'join'
+      ? 'https://realdevsquad.com/join'
       : 'https://realdevsquad.com';
 
   window.location.href = finalPage;
 }
 
+function redirectToJoinIfNotVisited() {
+  if (hasJoinVisited == 'true' || hasJoinVisited == null) {
+    redirectUserToPage('join.html');
```
Comment (ankushdharkar) [inline js/goto.js:16]: Can this just be `join` to maintain consistency? (Like `home` on line 18)

File: js/goto.js [hunk 2]
```diff
@@ -1,12 +1,24 @@
...
     page == 'signup'
       ? 'https://my.realdevsquad.com/signup?state=get-started'
+      : page == 'join'
+      ? 'https://realdevsquad.com/join'
       : 'https://realdevsquad.com';
 
   window.location.href = finalPage;
 }
 
+function redirectToJoinIfNotVisited() {
```
Comment (ankushdharkar) [inline js/goto.js:14]: Is this function name a bit confusing? I thought we wanted to go to `join` if we had previously visited it.

**Review by ankushdharkar — CHANGES_REQUESTED** [review]

File: js/goto.js (hunk 2 above)
Comment (sahsisunny) [reply js/goto.js]: changed to `redirectHandler()`.

File: js/goto.js [hunk 3]
```diff
@@ -1,27 +1,42 @@
...
-      : 'https://realdevsquad.com';
+    page == SIGNUP ? SIGNUP_URL : page == JOIN ? JOIN_URL : HOME_URL;
 
   window.location.href = finalPage;
 }
 
+function redirectionHandler(data) {
+  if (data.incompleteUserDetails) {
+    redirectUserToPage(SIGNUP);
+  } else if (hasJoinVisited == 'true' || hasJoinVisited == null) {
```
Comment (shmbajaj) [inline js/goto.js:22]: If I am not wrong, condition has to be 
```js
else if(hasJoinVisited !== 'true' || hasJoinVisited === null)
{}
```

File: js/goto.js [hunk 4]
```diff
@@ -1,27 +1,42 @@
...
   const finalPage =
-    page == 'signup'
-      ? 'https://my.realdevsquad.com/signup?state=get-started'
-      : 'https://realdevsquad.com';
+    page == SIGNUP ? SIGNUP_URL : page == JOIN ? JOIN_URL : HOME_URL;
 
   window.location.href = finalPage;
 }
 
+function redirectionHandler(data) {
```
Comment (shmbajaj) [inline js/goto.js:19]: `[Alternative]Suggestion`
```js
function redirectionHandler(data) {
  const hasVisitedJoin = hasJoinVisited == 'true' || hasJoinVisited == null;
  
  const nextPage = data?.incompleteUserDetails ? SIGNUP : hasVisitedJoin ? JOIN : HOME;
  
  redirectUserToPage(nextPage);
}
```

Comment (shmbajaj) [reply js/goto.js:19]:
```js
const PAGE_MAPPING = [
  {
    condition: data.incompleteUserDetails,
    page: SIGNUP,
  },
  {
    condition: hasJoinVisited === 'true' || hasJoinVisited === null,
    page: JOIN,
  },
];

function redirectionHandler(data) {
  let nextPage = HOME; // default page

  for (const { condition, page } of PAGE_MAPPING) {
    if (condition(data)) {
      nextPage = page;
      break;
    }
  }
  
  redirectUserToPage(nextPage);
}
```

Another Alternative!!

File: goto.html [hunk 5]
```diff
@@ -7,7 +7,7 @@
     <link rel="icon" href="./img/favicon.ico" type="image/x-icon" />
     <link rel="stylesheet" href="./css/main.css" />
 
-    <script type="text/javascript" src="js/goto.js"></script>
+    <script type="module" src="js/goto.js"></script>
```
Comment (RitikJaiswal75) [inline goto.html:10]: any specific reason for importing as module?

File: join.html [hunk 6]
```diff
@@ -394,12 +394,9 @@ <h2 class="progress">Here is your personalized link</h2>
           <ol>
             <li>Copy the above link</li>
             <li>
-              Follow Ankush Dharkar on
-              <a href="https://twitter.com/ankushdharkar">Twitter</a>
-            </li>
-            <li>
-              Message this link to him with a text of Real Dev Squad
-              Verification Link.
+              Message this link to Ankush on
```
Comment (RitikJaiswal75) [inline join.html:397]: can we change it to ->`Message this link to Ankush over direct message over Twitter`

File: js/constants.js [hunk 7]
```diff
@@ -261,3 +261,10 @@ export const JOIN_POST_URL = `${BASE_URL}/users/self/intro`;
 export const SELF_URL = `${BASE_URL}/users/self`;
 export const GITHUB_OAUTH =
```
Comment (RitikJaiswal75) [inline js/constants.js:262]: can we change it to `api.realdevsquad.com/auth/github/login ` with respect to the following commit: https://github.com/Real-Dev-Squad/website-backend/commit/c8f53da7970aea9685a50d7c2a622625b4aee881

File: js/constants.js [hunk 8]
```diff
@@ -261,3 +261,10 @@ export const JOIN_POST_URL = `${BASE_URL}/users/self/intro`;
...
   'https://github.com/login/oauth/authorize?client_id=23c78f66ab7964e5ef97';
+export const SIGNUP = 'signup';
+export const HOME = 'home';
+export const JOIN = 'join';
+export const SIGNUP_URL =
+  'https://my.realdevsquad.com/signup?state=get-started';
+export const JOIN_URL = 'https://realdevsquad.com/join';
```
Comment (RitikJaiswal75) [inline js/constants.js:269]: change it to ` ${HOME_URL}/${JOIN_URL}`

**Review by RitikJaiswal75 — CHANGES_REQUESTED** [review]

File: goto.html (hunk 5 above)
Comment (sahsisunny) [reply goto.html:10]: because i am importing some constants in the `goto.js` file.

File: join.html (hunk 6 above)
Comment (sahsisunny) [reply join.html:397]: i don't think it's needed.

File: js/goto.js (hunk 3 above)
Comment (sahsisunny) [reply js/goto.js:22]: If user go through the join page url then `hasJoinVisited == true` or user go through any site(Except join page url) then `hasJoinVisited == null`  then we redirected to join page section. After completed the join section `hasJoinVisited == false` then user redirect to home page.

File: js/goto.js (hunk 4 above)
Comment (sahsisunny) [reply js/goto.js:19]: this changes is necessary can we can't keep it simple ?

File: join.html (hunk 6 above)
Comment (RitikJaiswal75) [reply join.html]: nit: it is, try to ask other people on what could be the text here

File: js/constants.js [hunk 9]
```diff
@@ -261,3 +261,9 @@ export const JOIN_POST_URL = `${BASE_URL}/users/self/intro`;
 export const SELF_URL = `${BASE_URL}/users/self`;
 export const GITHUB_OAUTH =
   'https://github.com/login/oauth/authorize?client_id=23c78f66ab7964e5ef97';
+export const SIGNUP = 'signup';
+export const HOME = 'home';
```
Comment (RitikJaiswal75) [inline js/constants.js:265]: we have a HOME_URL, do we need these home and signup?

Comment (sahsisunny) [reply js/constants.js:265]: For yes for checking the conditions

File: join.html (hunk 6 above)
Comment (sahsisunny) [reply join.html]: text is good Ankush agrees

File: js/goto.js [hunk 10]
```diff
@@ -1,27 +1,38 @@
+import { SIGNUP, JOIN, HOME, SIGNUP_URL, HOME_URL } from './constants.js';
+
+const hasJoinVisited = localStorage.getItem('hasJoinVisited');
```
Comment (prakashchoudhary07) [inline js/goto.js:3]: NIT: `hasJoinVisited` -> `hasVisitedJoin`

File: js/join.js [hunk 11]
```diff
@@ -350,8 +351,9 @@ submit.addEventListener('click', async () => {
     .catch((err) => {
       alert(`Error in saving user data ${err}`);
     });
+  window.localStorage.setItem('hasJoinVisited', false);
```
Comment (prakashchoudhary07) [inline js/join.js:354]: When do we set and reset hasJoinVisited ?

**Review by prakashchoudhary07 — APPROVED** [review]

File: js/join.js (hunk 11 above)
Comment (sahsisunny) [reply js/join.js:354]: if user complete the join section

Comment (sahsisunny) [reply js/join.js:354]: for more info see [this](https://github.com/Real-Dev-Squad/website-www/pull/425#discussion_r1131811425)

File: js/goto.js [hunk 12]
```diff
@@ -1,27 +1,38 @@
...
   const finalPage =
-    page == 'signup'
-      ? 'https://my.realdevsquad.com/signup?state=get-started'
-      : 'https://realdevsquad.com';
-
+    page == SIGNUP
+      ? SIGNUP_URL
+      : page == JOIN
+      ? `${HOME_URL}/${JOIN}`
+      : HOME_URL;
```
[the remark this answers is not in the document]
Comment (sahsisunny) [reply js/goto.js]: done

File: js/goto.js (hunk 10 above)
Comment (sahsisunny) [reply js/goto.js]: done

**Review by satyam73 — CHANGES_REQUESTED** [review]

**Review by satyam73 — APPROVED** [review]
LGTM👍

**Pull request merged**